azure-pipelines.yml

```diff
@@ -1,56 +1,138 @@
-    summaryFileLocation: '$(Build.SourcesDirectory)/**/coverage.cobertura.xml'
+  - job: 'Build_Package'
+    displayName: Build Package
+    dependsOn: build_prerequisites
+    steps:
+    - task: NuGetToolInstaller@1
+      displayName: Install NuGet Tool
+    - task: Cache@2
+      displayName: 'NuGet Cache'
+      inputs:
+        key: 'nuget | "$(Agent.OS)" | **/packages.lock.json,!**/bin/**,!**/obj/**'
+        restoreKeys: |
+            nuget | "$(Agent.OS)"
+            nuget
+        path: $(NUGET_PACKAGES)
+        cacheHitVar: 'NUGET_CACHE_RESTORED'
+    - task: NuGetCommand@2
+      displayName: Nuget Restore
+      inputs:
+        restoreSolution: '$(projectNamePrefix).sln'
+    - task: DotNetCoreCLI@2
+      displayName: 'Build package'
+      inputs:
+        command: 'build'
+        projects: '$(ProjectNamePrefix).sln'
+        configuration: $(buildConfiguration)
+        arguments: /p:TreatWarningsAsErrors=true
+    - task: DotNetCoreCLI@2
+      displayName: Run Unit Tests
+      inputs:
+        command: 'test'
+        projects: $(projectNamePrefix)**/*.Tests.csproj
+        arguments: '--filter TestCategory=unit|TestCategory=smoke --configuration $(buildConfiguration)'
+        nobuild: true
+    # - task: DotNetCoreCLI@2
+    #   displayName: Pack the NuGet packages
+    #   inputs:
+    #     command: 'pack'
+    #     packagesToPack: '**/*.csproj'
+    #     configuration: '$(buildConfiguration)'
+    #     versioningScheme: 'byBuildNumber'
+    - task: DotNetCoreCLI@2
+      displayName: Pack the NuGet packages
+      inputs:
+        command: 'pack'
+        packagesToPack: '**/*.csproj'
+        configuration: '$(buildConfiguration)'
+        versioningScheme: 'byEnvVar'
+        versionEnvVar: 'BUILD_BUILDNUMBER'
+    - task: PublishBuildArtifacts@1
+      displayName: Publish artifacts
+      inputs:
+        PathtoPublish: '$(Build.ArtifactStagingDirectory)'
+        ArtifactName: $(artifactName)-$(Build.BuildNumber)
+        publishLocation: 'Container'
 
-- task: DotNetCoreCLI@2
-  displayName: 'Create NuGet Packages'
-  inputs:
-    command: 'pack'
-    packagesToPack: '$(ProjectNamePrefix).sln'
-    versioningScheme: 'byBuildNumber'
-    configuration: $(buildConfiguration)
+  - template: build-report-job.yml@templates
+    parameters:
+      artifactName: $(artifactName)
+      displayName: 'Generate Reports'
+      projectNameAngularApp: ''
+      projectNamePrefix: $(projectNamePrefix)
 
-- task: DotNetCoreCLI@2
-  displayName: 'Push NuGet Packages'
-  inputs:
-    command: 'push'
-    packagesToPush: '$(Build.ArtifactStagingDirectory)/*.nupkg'
-    nuGetFeedType: 'internal'
-    publishVstsFeed: 'b10976dd-2e80-4940-82ff-0f27c8e76baa/d86fc1c0-d5ab-4b0a-8b57-5dc7dd5a60cb'
\ No newline at end of file
+- stage: publish_nuget
+  displayName: publish packages to NuGet
+  dependsOn: ci_build
+  condition: and(succeeded(), startsWith(variables['Build.SourceBranch'], 'refs/heads/master'))
```

Comment:
I would suggest that instead of this condition we try with:

`condition: and(succeeded(), startsWith(variables['build.sourceBranch'], 'refs/tags/'))`

This will trigger publishing to NuGet only after we tag it with appropriate version (with or without 'preview').

We might need to trigger this run also on tags being pushed.

E.g.

`trigger:
  branches:
    include:
    - '*'
  tags:
    include: 
    - '*'`

Reply:
So if someone tags a feature branch (for whatever reason), it will also be published?

Reply:
I have never noticed anyone using tags for anything. If you think that we need to be more careful, then we can improve condition to check for tags that match certain regex. e.g. _v[0-9]+.[0-9]+.[0-9]+_ or _v[0-9]+.[0-9]+.[0-9]+-preview[0-9]+_